### ISSUE TITLE
Switch to Piston's image library

### ISSUE THREAD

**Issue by jdm** [servo/servo#3368]
[rust-image](https://github.com/PistonDevelopers/rust-image) appears to support everything we require from an image library right now, and it's native rust code. It would be nice to scrap stb-image and rust-png if possible, but we should definitely take some measurements to sure what the performance is like before and after switching.

**Comment by kmcallister** [comment]
Is there any update on this? @pcwalton says stb-image is the biggest blocker for Facebook timeline -- it can't decode any of their images. That's aside from the security issue that we need to resolve before dogfooding or a tech demo.

I'm not too worried about image decoding performance at the moment. It's tiny compared to some of the current inefficiencies in resource loading etc. If it shows up in profiles then we can investigate.  
Right now Servo's interaction with image decoders is very basic: send some bytes, get (width, height) and a vector of pixels. It's not hard to switch decoders, and Piston's library seems like a good medium-term solution.

**Comment by jdm** [comment]
https://github.com/servo/servo/pull/4215 worked, but the students never addressed the review comments and it needs a new owner.

**Comment by kmcallister** [comment]
I'll try to revive that PR.

**Comment by kmcallister** [comment]
Working on this again now that the rustup has landed.

**Comment by kmcallister** [comment]
Now waiting on PistonDevelopers/image#401.

**Comment by kmcallister** [comment]
And PistonDevelopers/image#403.

**Comment by kmcallister** [comment]
Those are fixed; now it's PistonDevelopers/image#409.

**Comment by waddlesplash** [comment]
That was fixed in PistonDevelopers/image#412, so there are no more blockers, correct?

**Comment by kmcallister** [comment]
There are some other issues I didn't file yet :/

I hit enough bugs in a row on our testsuite (which doesn

**Comment by kmcallister** [comment]
Now tracking these as PistonDevelopers/image#413.

**Comment by frewsxcv** [comment]
@notriddle Did you make progress on this? Run into any issues? I noticed you have a commit on your forked Servo

**Comment by notriddle** [comment]
I had it mostly working (WebP wouldn't work, but I think that might be a
problem in Piston-Image). But then they made images serializable, and that
broke it.

**Comment by jdm** [comment]
@notriddle Other libraries have accepted configurable serialization PRs, such as https://github.com/servo/rust-url/pull/118 and https://github.com/hyperium/hyper/pull/603 . Might be worth a try!

**Comment by paulrouget** [comment]
We need this to decode favicons in browser.html. Do we still want to use this library?

**Comment by metajack** [comment]
We would need to use some favicon library. stb-image is not usable long term (if it supports favicons and that is your concern). Maybe we need to add favicon support to rust-image?

**Comment by mbrubeck** [comment]
Now that images use shared memory (#6705), just adding a serde feature to the `image` crate is not enough; we'd need to make it use the `ipc-channel` crate too and change how it allocates memory for image data.

Rather than push all this stuff directly into `image`, I think in the short term we can should use `image` similarly to how we currently use `png` and `stb_image`: Just use it to decode the data, then copy the results into our shared-memory image struct.

Compared to the code in #4215 this will mean keeping more of Servo's current code around, at least for now.  I think I'll start a new PR from scratch for this short-term solution, unless there are any objections.

(In the long term when Rust's "placement new" equivalent is fully implemented and stabilized, maybe we can make `image` work with generic custom "placers" including one for IpcSharedMemory, and then revive the changes from #4215 that uses `image` types more directly.)

**Comment by frewsxcv** [comment]
For what it's worth, this PR caused a regression and Google News now panics: https://github.com/servo/servo/issues/8093
